source/game/kart/KartMove.cc

```diff
@@ -129,67 +170,266 @@ void KartMove::calc() {
+
+        f32 fVar4 = 1.0f;
+        if (cross.dot(m_smoothedUp) < 0.0f) {
+            fVar4 = -1.0f;
         }
+        m_landingAngle += (angle * RAD2DEG) * fVar4;
     }
 
-    m_vel1Dir = m_dir.perpInPlane(m_smoothedUp, true);
+    if (m_landingAngle <= 0.0f) {
+        if (m_landingAngle < 0.0f) {
+            m_landingAngle = std::min(0.0f, m_landingAngle + 2.0f);
+        }
+    } else {
+        m_landingAngle = std::max(0.0f, m_landingAngle - 2.0f);
+    }
 }
 
 void KartMove::calcOffroad() {
-    m_kclRotFactor = m_kclWheelRotFactor / static_cast<f32>(m_floorCollisionCount);
+    if (state()->isAnyWheelCollision()) {
+        m_kclSpeedFactor = m_kclWheelSpeedFactor;
+        m_floorCollisionCount = m_floorCollisionCount != 0 ? m_floorCollisionCount : 1;
+        m_kclRotFactor = m_kclWheelRotFactor / static_cast<f32>(m_floorCollisionCount);
+    }
+}
+
+bool KartMove::calcPreDrift() {
+    if (!state()->isTouchingGround() && !state()->isHop() && !state()->isDriftManual()) {
+        if (state()->isStickLeft() || state()->isStickRight()) {
+            if (m_hopStickX == 0) {
+                if (state()->isStickRight()) {
+                    m_hopStickX = -1;
+                } else if (state()->isStickLeft()) {
+                    m_hopStickX = 1;
+                }
+                cancelWheelie();
+            }
+        }
+    }
+
+    if (state()->isHop()) {
+        if (m_hopStickX == 0) {
+            if (state()->isStickRight()) {
+                m_hopStickX = -1;
+            } else if (state()->isStickLeft()) {
+                m_hopStickX = 1;
+            }
+        }
+        if (m_hopFrame < 3) {
+            ++m_hopFrame;
+        }
+    } else {
+        m_hopFrame = 0;
+    }
+
+    return state()->isHop();
+}
+
+void KartMove::resetDriftManual() {
+    m_hopStickX = 0;
+    m_hopFrame = 0;
+    state()->setHop(false);
+    state()->setDriftManual(false);
+    m_driftState = DriftState::NotDrifting;
+    m_mtCharge = 0;
+}
+
+void KartMove::calcManualDrift() {
+    bool hop = calcPreDrift();
+
+    // TODO: Is this backwards/inverted?
+    if ((!state()->isHop() || m_hopFrame < 3) || !state()->isTouchingGround()) {
+        if (canHop()) {
+            calcHop();
+            hop = true;
+        }
+    } else {
+        startManualDrift();
+        hop = false;
+    }
+
+    if (!state()->isDriftManual()) {
+        if (!hop && state()->isTouchingGround()) {
+            resetDriftManual();
+        }
+    } else {
+        if (!state()->isDriftInput() || !state()->isAccelerate()) {
+            releaseMt();
+            resetDriftManual();
+        } else {
+            controlOutsideDriftAngle();
+        }
+    }
+}
+
+void KartMove::startManualDrift() {
+    constexpr f32 MIN_DRIFT_THRESHOLD = 0.55f;
+
+    state()->setHop(false);
+
+    if (!state()->isDriftInput()) {
+        return;
+    }
+
+    if (MIN_DRIFT_THRESHOLD * m_baseSpeed > m_speed) {
+        return;
+    }
+
+    if (m_hopStickX != 0) {
```

Comment:
See 8057ea94 and 8057eff8. This simplifies to:
```suggestion
    if (getAppliedHopStickX() == 0) {
        return;
    }
```